easybuild/tools/filetools.py

```diff
@@ -2707,11 +2707,16 @@ def get_source_tarball_from_git(filename, targetdir, git_config):
             for cmd in cmds:
                 run.run_cmd(cmd, log_all=True, simple=True, regexp=False, path=repo_name)
 
-    # create an archive and delete the git repo directory
-    if keep_git_dir:
-        tar_cmd = ['tar', 'cfvz', targetpath, repo_name]
-    else:
-        tar_cmd = ['tar', 'cfvz', targetpath, '--exclude', '.git', repo_name]
+    # When CentOS 7 is phased out and tar>1.28 is everywhere, replace find-sort-pipe with tar-flag
+    # '--sort=name' and place LC_ALL in front of tar. Also remove flags --null, --no-recursion, and
+    # --files-from - from the flags to tar. See https://reproducible-builds.org/docs/archives/
+    tar_cmd = ['find', repo_name, '-print0', '|',
+               'LC_ALL=C', 'sort', '--zero-terminated', '|',
+               'GZIP=--no-name', 'tar', '--create', '--file', targetpath, '--gzip',
+               '--mtime="1970-01-01 00:00Z"', '--owner=0', '--group=0', '--numeric-owner',
+               '--format=gnu', '--null',
+               '--exclude=.git' if not keep_git_dir else '',
+               '--no-recursion', '--files-from -']
```

Comment:
This looks well crafted. I would prefer to have this separate from the other change as it would apply independently.

Is this stable enough with possibly changing versions of gzip as well? We could also opt to not gzip at all if that would help reproducibility across systems.

Reply:
It should be compatible with newer versions of GNU Gzip and Tar. It is incompatible with for example FreeBSD's Tar, based on a cursory look at its manpage which does not mention a `--mtime` flag.

Dropping the gzipping of the files would not help a lot; given that all changes but the flag `--no-name` are to Tar and not Gzip.

Reply:
Although this approach could clearly help, but I have a couple concerns:

1) This may not be enough to get 100% reproducible tarballs across a variety of systems. I'm fairly sure that a different version of `gzip` being used may already be sufficient, since that may do better compression, etc. Another factor could be locale, which may affect the order of paths produced with the `find` command.
   If the tarball creation not 100% reproducible across a wide variety of systems, it's basically useless, since people who get a slightly different tarball will be running into incorrect checksums all the time, which would not be good.
2) The side effect of using `--mtime` to reset time stamps and `--owner=0` is that information gets lost. Especially the timestamp info may be interesting, so it should be retained.
3) This likely breaks the creating of tarball for easyconfigs using `git_config` on macOS, which is a regression. Although you currently won't get very far with installing software on macOS with EasyBuild, features like downloading sources with `eb --fetch` should work.

I applaud your effort here @Rovanion, but I'm not sure it's wise to go forward with this approach...

Reply:
Changing mtime is an absolute do-not-do-that.

It also has the problem of not generating a tar file that has the same content as what tar zcf does.
A small test I did shows that the file/sort/tar pipe above actually adds each file twice in the tar file.

Example:
```
mkdir q
echo hej > q/t
echo hopp > q/w
find q | sort | tar --create --file q.tar --files-from -
tar cf q2.tar q
tar tf q2.tar 
q/
q/t
q/w
tar tf q.tar 
q/
q/t
q/w
q/t
q/w
```
this is caused by directory coming first, thereby making tar copy the whole dir into the tar file, then the list of files, making tar add the files once more.